docs/allocation.md

```diff
@@ -0,0 +1,152 @@
+* availability set forth in the Eclipse Public License, v. 2.0
+* are satisfied: GNU General Public License, version 2 with
+* the GNU Classpath Exception [1] and GNU General Public
+* License, version 2 with the OpenJDK Assembly Exception [2].
+*
+* [1] https://www.gnu.org/software/classpath/license.html
+* [2] http://openjdk.java.net/legal/assembly-exception.html
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH
+* Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+
+# Heap allocation
+
+The process of managing memory in the VM is handled by the allocator and the garbage collector. These components operate on an area of memory that is reserved for VM processing called the Java&trade; heap.
+
+The allocator assigns areas of the heap for Java objects. Objects are considered as *live* when they have a chain of references to them that start from root references, such as those found in thread stacks. When that reference or pointer no longer exists, the objects are considered as *garbage*.
+
+The garbage collector reclaims memory by removing objects when they are no longer required. To find out more about the garbage collector, see [Garbage collection](gc_overview.md).
+
+Depending on your application workload or service level agreement, you can choose from a number of OpenJ9 *garbage collection (GC) policies*. Each GC policy uses a different strategy to manage memory on the heap. The structure of the heap also depends on the strategy in force. For more information about choosing a GC policy, see [Garbage collection policies](gc.md).
+
+## The allocator
+
+The allocator manages pools of free memory and how the free memory is consumed. It is also responsible for allocating areas of storage in the Java heap for objects at the request of applications, class libraries, or the VM.
+
+Every allocation requires a *heap lock* to stop concurrent threads trying to access the same area of memory at the same time. When an object is allocated, the heap lock is released. If there is insufficient space to allocate the object, allocation fails, the heap lock is released, and the GC is called. If the GC manages to recover some space on the heap, the allocator can resume operations. If the GC does not recover enough space, it returns an `OutOfMemoryError` exception.
+
+Acquiring a heap lock for every allocation would be an intensive operation with a knock on impact to performance. To get around this problem, small objects are allocated to allocation caches.
+
+### Allocation caches
+
+To improve performance, allocation caches are reserved in the heap for different threads. These allocation caches are known as thread local heaps (TLH) and allow each thread to allocate memory from its cache without acquiring the heap lock. A TLH is typically used for small objects of less than 512 bytes (768 bytes on 64-bit VMs) although larger objects can be allocated from the cache if there is sufficient space.
```

Comment:
This only partially true: 'A TLH is typically used for small objects of less than 512 bytes (768 bytes on 64-bit VMs) although larger objects can be allocated from the cache if there is sufficient space.'

This is more correct (and partially covers the aspect of TLH refresh):

Objects are attempted to allocate from TLH regardless of its size. If TLH cannot satisfy the allocation, depending on the size left in TLH (and some internal heuristics), and the size of the allocation, the allocation may proceed as individual allocation directly from the heap (typical for larger objects, certainly true for objects larger than TLH max) or we may proceed to refresh TLH (discard the existing one and create new one) and allocate from the new TLH.

Reply:
Adapted, which I hope gives the user enough. If not, I can expand:

"Objects are allocated from the TLH unless there is insufficient space remaining in the TLH to satisfy the request. In this situation, the allocation might proceed directly from the heap by using a heap lock or the TLH might be refreshed to increase the size of the cache."

Reply:
' to increase the size of the cache.' is not quite correct and probably can be dropped altogether (if you are already trying to keep it simple). If we refresh TLH it's just simply to get a new empty TLH (to be able to satisfy ongoing allocation request). TLH increase is just a secondary effect that may occur, if we have not reached TLH max size already. If we did reach, the refresh will occur with that same TLH size as before.

I'd still consider adding 'for larger objects' somewhere  for the case of 'the allocation might proceed directly from the heap by using a heap lock' (what would imply that TLH will be refreshed for the other case of smaller objects).